source/code-snippets/transactions/txn-core.js

```diff
@@ -0,0 +1,137 @@
+  await Promise.all( ['customers', 'inventory', 'orders'].map(async c => {
+    try {
+      const coll = client.db('testdb').collection(c);
+      await coll.drop();
+    } catch(e) {}
+  }));
+}
+
+async function setup(client) {
+  try {
+    const customerColl = client.db('testdb').collection('customers');
+    const inventoryColl = client.db('testdb').collection('inventory');
+
+    await customerColl.insertOne({ _id: 98765, orders: [] });
+
+    await inventoryColl.insertMany([
+      { name: 'sunblock', sku: 5432, qty: 85 },
+      { name: 'beach towel', sku: 7865, qty: 41 },
+    ]);
+  } catch (e) {
+    console.log('Unable to insert test data: ' + e);
+  }
+}
+
+async function queryData() {
+  const uri = process.env.MONGODB_URI;
+  const client = new MongoClient(uri, { useUnifiedTopology: true });
+  try {
+    await client.connect();
+    await Promise.all(['customers', 'inventory', 'orders'].map(async c => {
+      const coll = client.db('testdb').collection(c);
+      console.log(JSON.stringify(await coll.find().toArray()));
+    }, client));
+
+  } finally {
+    client.close();
+  }
+}
+
+// start placeOrder
+async function placeOrder(client, cart, payment) {
+  const transactionOptions = {
+    readConcern: { level: 'snapshot' },
+    writeConcern: { w: 'majority' },
+    readPreference: 'primary'
+  };
+
+  const session = client.startSession();
+  try {
+    session.startTransaction(transactionOptions);
+
+    const ordersCollection = client.db('testdb').collection('orders');
+    const orderResult = await ordersCollection.insertOne(
+      {
+        customer: payment.customer,
+        items: cart,
+        total: payment.total,
+      },
+      { session }
+    );
+
+    const inventoryCollection = client.db('testdb').collection('inventory');
+    for (var i = 0; i < cart.length; i++) {
+      const item = cart[i];
+
+      // Cancel the transaction when you have insufficient inventory
+      const checkInventory = await inventoryCollection.findOne(
+        {
+          sku: item.sku,
+          qty: { $gte: item.qty }
+        },
+        { session }
+      )
+      if (checkInventory === null) {
+        throw new Error('Insufficient quantity or SKU not found.');
+      }
+
+      await inventoryCollection.updateOne(
+        { sku: item.sku },
+        { $inc: { 'qty': -item.qty }},
+        { session }
+      );
+    }
+
+    const customerCollection = client.db('testdb').collection('customers');
+    await customerCollection.updateOne(
+      { _id: payment.customer },
+      { $push:  { orders: orderResult.insertedId }},
+      { session }
+    );
+    await session.commitTransaction();
+    console.log('Transaction successfully committed.');
+
+  } catch (error) {
+    if (error instanceof MongoError) {
+      if (err.hasErrorLabel('UnknownTransactionCommitResult')) {
+        // add your logic to retry or handle the error
+      }
+      else if (err.hasErrorLabel('TransientTransactionError')) {
```

Comment:
[comment, maybe an issue]

I realize the snippet that appears on the page is only part of this larger example, but when pulling just that part of it and running it on my own, I run into:
- `MongoError` is not defined (line 99), and then
- `err` is not defined (lines 100, 103)

I solved this by removing the `instanceof MongoError` check (now just `(if error)`) and changing `err` to `error`.

Reply:
Whoops, good catch! Not immediately sure why it didn't error out when I ran it.

Reply:
Changed this around a bit to ensure the reader understands to import MongoError and to pair the check that it's a MongoError (to guarantee that there's a function on it "hasErrorLabel").